torchgeo/datasets/utils.py

```diff
@@ -51,9 +56,108 @@
     "draw_semantic_segmentation_masks",
     "rgb_to_mask",
     "percentile_normalization",
+    "PredictDataset",
 )
 
 
+class PredictDataset(Dataset[Any]):
```

Comment:
Can you clarify why this is necessary?

Reply:
`PredictDataset` represents a dataset of images in a folder that you want to run inference on. When you run `predict.py` you need to pass a `predict_on` flag. This can be either a test split of a dataset or a folder of images that you want to run inference on. `PredictDataset` handles the latter. 

It accepts a bands argument since we can't predict the number of bands the geotiffs will have and one transform (`patch_sample`) that it will always do since we can't predict the size of the geotiff.

Reply:
Hmm, I'm still not understanding. Why not use the normal data loader for that dataset (a subclass of either `VisionDataset` or `GeoDataset`)? `RasterDataset` in combination with `GridGeoSampler` (or just a normal `VisionDataset`) is specifically designed to handle "a folder of images you want to run inference on". It seems like this won't work for all datasets since it only handles flat directories of images that can be opened with rasterio and don't need to be chipped.

Reply:
> Why not use the normal data loader for that dataset (a subclass of either `VisionDataset` or `GeoDataset`)? 

Because the dataloader of that dataset has custom logic for parsing filenames and subdirectories that is not relevant when you want to run inference on a directory of images. You could add code to the `_load_files` method of each dataset that skips the parsing logic if it is in "predict" mode and I could see a case being made for that.

I just found having a separate class (PredictDataset) that represents a folder of images to be easier. Just need to instantiate it in the corresponding datamodule as seen in this PR. 


> It seems like this won't work for all datasets since it only handles flat directories of images that can be opened with rasterio and don't need to be chipped.

That's the exact use case being targeted. You have a bunch of images of arbitrary size in a folder and you want to produce inference results.  

These images will of course need to be chipped to the size the model expects which is why the `PredictDataset` contains a single transform that it will **always** apply - `patch_sample` which extracts tensor patches according to the `patch_size` the model was trained on.

Reply:
> Because the dataloader of that dataset has custom logic for parsing filenames and subdirectories that is not relevant when you want to run inference on a directory of images.

But this doesn't cause any issues, right? Yes, you could use a simpler class for parsing a directory of images, but we already have datasets for everything. If there's no reason why those existing datasets don't work then I don't see why we need a new dataset.

Reply:
This will cause issues. A dataset's default `load_files` is dependent on how that particular dataset is organised whereas our inference folder is just a folder of images. Consider the `load_files` method of the Inria dataset:

```
    def _load_files(self, root: str) -> List[Dict[str, str]]:
        """Return the paths of the files in the dataset.

        Args:
            root: root dir of dataset

        Returns:
            list of dicts containing paths for each pair of image and label
        """
        files = []
        root_dir = os.path.join(root, self.directory, self.split)
        images = glob.glob(os.path.join(root_dir, "images", "*.tif"))
        images = sorted(images)
        if self.split == "train":
            labels = glob.glob(os.path.join(root_dir, "gt", "*.tif"))
            labels = sorted(labels)

            for img, lbl in zip(images, labels):
                files.append({"image": img, "label": lbl})
        else:
            for img in images:
                files.append({"image": img})

        return files
```

This was written to address the following dataset organisation:

```
inria/
└── AerialImageDataset
    ├── test
    │   └── images
    │       ├── bellingham1.tif
    │       └── bellingham2.tif
    └── train
        ├── gt
        │   ├── austin1.tif
        │   ├── austin2.tif
        │   ├── austin3.tif
        └── images
            ├── austin1.tif
            ├── austin2.tif
            ├── austin3.tif

```

whereas the inference folder is just a folder of images

```
inference/
├── img1.tif
└── img2.tif
└── img3.tif
```


Now as I mentioned before we could reuse the existing dataset and not use PredictDataset. For that you would need to account for this in the `load_files` function as follows:

```
    def _load_files(self, root: str) -> List[Dict[str, str]]:
        """Return the paths of the files in the dataset.

        Args:
            root: root dir of dataset

        Returns:
            list of dicts containing paths for each pair of image and label
        """
        files = []
        # An attribute to indicate we are running inference on a folder of images. 
        if self.predict:
            images = [os.path.join(root, i) for i in os.listdir(root)]
            return [{"image": img} for img in images]
       else:
            root_dir = os.path.join(root, self.directory, self.split)
            images = glob.glob(os.path.join(root_dir, "images", "*.tif"))
            images = sorted(images)
            if self.split == "train":
                labels = glob.glob(os.path.join(root_dir, "gt", "*.tif"))
                labels = sorted(labels)
    
                for img, lbl in zip(images, labels):
                    files.append({"image": img, "label": lbl})
            else:
                for img in images:
                    files.append({"image": img})

            return files
```

So why didn't I go with this? Because this meant I would need to similarly modify the `load_files` of every dataset that I wanted to use to run inference. Having PredictDataset allows the separation of this prediction logic, is reusable and testable.

We generally have **two** types of prediction:

1. You want to predict on an existing test set of a dataset. 
2. You want to use a model trained on a dataset to predict on some general images.

For (1), you would use the existing dataset class whereas for (2) you would use `PredictDataset` as you can see in the inria datamodule. That was the idea.

Reply:
Okay, I think I'm starting to understand. It sounds like you want to train a model on Inria and then make predictions on a completely different dataset with _similar_ images. That's fine, but maybe a name other than `PredictDataset` would be better. I was under the mistaken impression that `PredictDataset` would be required for all datasets you want to use in `predict.py`.

We already have `RasterDataset` and `VectorDataset` subclasses of `GeoDataset` that recursively locate files that can be opened with rasterio and fiona. Maybe we need a new `ImageDataset` subclass of `VisionDataset` that recursively locates files that can be opened with PIL? Or maybe that name would be too generic since most `VisionDataset`s contain images but can't subclass from `ImageDataset` because it's not possible to automatically find labels. We do have `VisionClassificationDataset` too. Btw, `VisionDataset` is going to be renamed to `NonGeoDataset` in the near future.

Reply:
> It sounds like you want to train a model on Inria and then make predictions on a completely different dataset with similar images

Precisely. For example, one of the locations within the inria dataset is Austin. So if someone gives me some aerial imagery with similar resolution (30cm) over Dallas, Houston or even Austin, I want to be able to use this model to run inference. 

>  That's fine, but maybe a name other than PredictDataset would be better.

Maybe `PredictionFolder`? I'm open to suggestions.

> I was under the mistaken impression that PredictDataset would be required for all datasets you want to use in predict.py.

Ah I see. `PredictDataset` was meant to be a standalone class to be used when you want to run inference on a folder of images of arbitrary size and no. of bands. You would inject it into the datamodule so that it can be used with `predict.py` as seen in this PR. It's not connected to any dataset. 


> We already have RasterDataset and VectorDataset subclasses of GeoDataset that recursively locate files that can be opened with rasterio and fiona. Maybe we need a new ImageDataset subclass of VisionDataset that recursively locates files that can be opened with PIL?

Do we need PIL? I think we can stick with `rasterio` but not use `RasterDataset`. Geospatial info is not necessary for the prediction process but we don't want the process to fail if the image lacks geospatial info. Apart from throwing a `NotGeoreferencedWarning`, `rasterio` can read and write any image with/without no geospatial info.

Reply:
Ah, didn't even notice that you were using rasterio. In that case, this class definitely isn't needed. You should definitely use `RasterDataset` for this specific dataset. You might not need geospatial info if you only want to predict on one file, but as soon as you want to predict on multiple files, you need geospatial info to stitch together the predictions.

Reply:
If the file in question does not have geospatial info then `RasterDataset` can't be used. Also the patching and merging/stitching are done by kornia's `ExtractTensorPatches` and `CombineTensorPatches`.

Just so we're on the same page this is how I visualise the workflow:

The main inputs to the `predict.py` script are a directory of images (inference folder) and the trained model. **The directory of images contain images with/without georeference and with different sizes**.

The script first runs over each image, patches/chips them to the appropriate size for the model, runs model inference and merges them back together.

Reply:
This works fine if you want to convert a directory of images to a directory of predictions, but what if you want to predict a single static map of an entire country by stitching together those predictions? I think that's where `predict.py` would really shine.

I still think the best approach is to use `RasterDataset` for geospatial data and some kind of subclass of `VisionDataset` for non-geospatial data. Also, these datasets could be used at training time for some kind of self-supervised learning, so `PredictDataset` wouldn't be the best name.

Reply:
It seems we're coming at this from two different angles. 

From my POV, `predict.py` was **only** intended to convert a directory of images to a directory of predictions. The motivation was simple: I have a model and I have images from certain areas of interest that I want to run inference on. I might want to merge them together but I could just use `gdal_merge.py` for that. 

The images could be spatially distributed i.e. `AOI1.tif`, `AOI2.tif` etc or temporally distributed i.e. `AOI_2020.tif`, `AOI_2019.tif`, `AOI_2018.tif`. 


Another place where this would be useful is when you want to upload predictions to an evaluation server. In Inria, the images are of size 5000 x 5000. For training I used 512 x 512 crops. But if I want to upload the predictions to an evaluation server they would expect predictions to be the same size as input i.e. 5000 x 5000.


Predicting a single static map or using it for self supervised learning is well outside my original scope.

In order to properly address this, there needs to some consensus on what prediction scenarios `torchgeo` is interested in supporting:

1. Will it support images with and without georeference?
2. Will it support running on a directory of images and producing a directory of predictions?
3. Will it provide an option to merge the directory of predictions if they have geo-reference into a single static map?